File: sdk/python/payraider/http.py
"""HTTP client with auth, retry, and rate-limit handling."""
from __future__ import annotations

import asyncio
import random
import time
from typing import Any, Optional

import httpx

DEFAULT_BASE_URL = "https://api.payraider.io"
DEFAULT_MAX_RETRIES = 3
DEFAULT_RETRY_DELAY = 0.5  # seconds
DEFAULT_TIMEOUT = 30.0  # seconds

RETRYABLE_STATUSES = {429, 500, 502, 503, 504}


class PayRaiderError(Exception):
    """Raised when the API returns an error response."""

    def __init__(self, status: int, code: str, message: str, request_id: Optional[str] = None) -> None:
        super().__init__(message)
        self.status = status
        self.code = code
        self.request_id = request_id

    def __repr__(self) -> str:
        return f"PayRaiderError(status={self.status}, code={self.code!r}, message={str(self)!r})"


class HttpClient:
    def __init__(
        self,
        *,
        api_key: Optional[str] = None,
        access_token: Optional[str] = None,
        base_url: str = DEFAULT_BASE_URL,
        max_retries: int = DEFAULT_MAX_RETRIES,
        retry_delay: float = DEFAULT_RETRY_DELAY,
        timeout: float = DEFAULT_TIMEOUT,
    ) -> None:
        self._base_url = base_url.rstrip("/")
        self._max_retries = max_retries
        self._retry_delay = retry_delay
        self._token: Optional[str] = access_token or api_key
        self._client = httpx.AsyncClient(timeout=timeout)

    def set_token(self, token: str) -> None:
        self._token = token

    def _headers(self) -> dict[str, str]:
        headers = {"Accept": "application/json", "Content-Type": "application/json"}
        if self._token:
            headers["Authorization"] = f"Bearer {self._token}"
        return headers

    async def request(
        self,
        method: str,
        path: str,
        *,
        params: Optional[dict[str, Any]] = None,
        json: Optional[Any] = None,
    ) -> Any:
        url = f"{self._base_url}{path}"
        # Strip None values from params
        clean_params = {k: v for k, v in (params or {}).items() if v is not None}

        for attempt in range(self._max_retries + 1):
            response = await self._client.request(
                method,
                url,
                headers=self._headers(),
                params=clean_params or None,
                json=json,
            )

            if response.status_code in RETRYABLE_STATUSES and attempt < self._max_retries:
                delay = self._backoff(response, attempt)
                await asyncio.sleep(delay)
                continue

            if not response.is_success:
                self._raise_error(response)

            if response.status_code == 204:
                return None
            return response.json()

        # Should not reach here, but raise last error
        self._raise_error(response)  # type: ignore[possibly-undefined]

    def _backoff(self, response: httpx.Response, attempt: int) -> float:
        retry_after = response.headers.get("Retry-After")
        if retry_after and retry_after.isdigit():
            return float(retry_after)
        return self._retry_delay * (2**attempt) + random.uniform(0, 0.1)

    def _raise_error(self, response: httpx.Response) -> None:
        try:
            body = response.json()
        except Exception:
            body = {}
        raise PayRaiderError(
            status=response.status_code,
            code=body.get("error", "UNKNOWN_ERROR"),
            message=body.get("message", response.reason_phrase or "Unknown error"),
            request_id=body.get("request_id"),
        )

    async def aclose(self) -> None:
        await self._client.aclose()

    async def __aenter__(self) -> "HttpClient":
        return self

    async def __aexit__(self, *_: Any) -> None:
        await self.aclose()

File: sdk/typescript/src/http.ts
import type { VeltroConfig, ApiError } from "./types.js";

const DEFAULT_BASE_URL = "https://api.veltro.io";
const DEFAULT_MAX_RETRIES = 3;
const DEFAULT_RETRY_DELAY = 500;
const DEFAULT_TIMEOUT = 30_000;

/** Errors that should trigger a retry */
const RETRYABLE_STATUSES = new Set([429, 500, 502, 503, 504]);

export class VeltroError extends Error {
  constructor(
    public readonly status: number,
    public readonly code: string,
    message: string,
    public readonly requestId?: string,
  ) {
    super(message);
    this.name = "VeltroError";
  }
}

export class HttpClient {
  private readonly baseUrl: string;
  private readonly maxRetries: number;
  private readonly retryDelay: number;
  private readonly timeout: number;
  private token: string | undefined;

  constructor(config: VeltroConfig) {
    this.baseUrl = (config.baseUrl ?? DEFAULT_BASE_URL).replace(/\/$/, "");
    this.maxRetries = config.maxRetries ?? DEFAULT_MAX_RETRIES;
    this.retryDelay = config.retryDelay ?? DEFAULT_RETRY_DELAY;
    this.timeout = config.timeout ?? DEFAULT_TIMEOUT;
    this.token = config.accessToken ?? config.apiKey;
  }

  /** Update the bearer token (e.g. after OAuth login) */
  setToken(token: string): void {
    this.token = token;
  }

  async request<T>(
    method: string,
    path: string,
    options: { body?: unknown; params?: Record<string, unknown>; headers?: Record<string, string> } = {},
  ): Promise<T> {
    const url = this.buildUrl(path, options.params);
    const headers: Record<string, string> = {
      "Content-Type": "application/json",
      Accept: "application/json",
    };
    if (this.token) {
      headers["Authorization"] = `Bearer ${this.token}`;
    }
    if (options.headers) {
      Object.assign(headers, options.headers);
    }

    const init: RequestInit = {
      method,
      headers,
      signal: AbortSignal.timeout(this.timeout),
    };
    if (options.body !== undefined) {
      init.body = JSON.stringify(options.body);
    }

    return this.withRetry(url, init);
  }

  private buildUrl(path: string, params?: Record<string, unknown>): string {
    const url = new URL(`${this.baseUrl}${path}`);
    if (params) {
      for (const [k, v] of Object.entries(params)) {
        if (v !== undefined && v !== null) {
          url.searchParams.set(k, String(v));
        }
      }
    }
    return url.toString();
  }

  private async withRetry<T>(url: string, init: RequestInit, attempt = 0): Promise<T> {
    const res = await fetch(url, init);

    if (!res.ok) {
      if (RETRYABLE_STATUSES.has(res.status) && attempt < this.maxRetries) {
        const retryAfter = this.getRetryAfter(res, attempt);
        await sleep(retryAfter);
        return this.withRetry<T>(url, init, attempt + 1);
      }
      await this.throwApiError(res);
    }

    // 204 No Content
    if (res.status === 204) return undefined as T;
    return res.json() as Promise<T>;
  }

  private getRetryAfter(res: Response, attempt: number): number {
    const header = res.headers.get("Retry-After");
    if (header) {
      const seconds = parseInt(header, 10);
      if (!isNaN(seconds)) return seconds * 1000;
    }
    // Exponential backoff with jitter
    return this.retryDelay * 2 ** attempt + Math.random() * 100;
  }

  private async throwApiError(res: Response): Promise<never> {
    let body: Partial<ApiError> = {};
    try {
      body = await res.json();
    } catch {
      // ignore parse errors
    }
    throw new VeltroError(
      res.status,
      body.error ?? "UNKNOWN_ERROR",
      body.message ?? res.statusText,
      body.request_id,
    );
  }
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}
